Binding for local deconstruction declaration statement

jcouv opened the pull request:
This change adds binding for local deconstruction declaration statements. The for and foreach statements will be added later.
The binding produces a `BoundExpressionStatement` with a `BoundDeconstructionAssignmentOperator`. The main difference compared to the binding for deconstruction-assignment is that now, variables in the `BoundDeconstructionAssignmentOperator` are `BoundLocal`.

Much of the binding still relies on the existing `BindDeconstructionAssignment`, except how the variables on the left are prepared. Going into `BindDeconstructionAssignment`, the variables are a new type, `DeconstructionLocalPendingInference`, which during the process of binding the assignment (once the types on the right-hand-side are figured out) get converted to `BoundLocal` with the inferred type. 
Like `OutVarLocalPendingInference`, `DeconstructionLocalPendingInference` doesn't survive the early binding process.
In order to allow the tree of variables to be updated by inference, I changed the existing `DeconstructionVariable` to hold an `ArrayBuilder` instead of `ImmutableArray` for nested elements.

@dotnet/roslyn-compiler for review.

gafter commented:
> This change adds binding for local deconstruction declaration statements. The for and foreach statements will be added later.
The binding produces a `BoundExpressionStatement` with a `BoundDeconstructionAssignmentOperator`. The main difference compared to the binding for deconstruction-assignment is that now, variables in the `BoundDeconstructionAssignmentOperator` are `BoundLocal`.

This approach does not match the design guidelines for initial bound nodes. The shape of the initial bound tree is intended to be fairly closely isomorphic to the shape of the syntax. We have separate bound nodes for statements and expressions, and the local declarations remain in the bound tree after initial binding. The deconstruction declaration is not an expression form in the language. It is represented in the syntax as something more akin to a local variable declaration (a statement).

It could be *translated* "via" an expression form during lowering, but it is not syntactically or semantically an expression statement.

jcouv commented:
@gafter Thanks for pointing this out. I will introduce a new bound node for `LocalDeconstructionStatement`. Do you see any concerns if that new node simply packages a `BoundDeconstructionAssignmentOperator` (rather than replicating its fields)?

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -23,12 +24,18 @@ internal partial class Binder
-
             SeparatedSyntaxList<ArgumentSyntax> arguments = ((TupleExpressionSyntax)node.Left).Arguments;
-            ImmutableArray<DeconstructionVariable> checkedVariables = BindDeconstructionVariables(arguments, diagnostics);
+            ArrayBuilder<DeconstructionVariable> checkedVariables = BindDeconstructionVariables(arguments, diagnostics);
```
Is `BindDeconstructionAssignment` method below freeing `checkedVariables`? Please consider having the same method allocate and free.

jcouv replied:
Thanks for pointing this. I think when I switched to `ArrayBuilder` I removed the freeing and did not put it back. I'll fix.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -41,20 +48,25 @@ private BoundExpression BindDeconstructionAssignment(AssignmentExpressionSyntax
+                FailRemainingInferences(checkedVariables, diagnostics);
                 return new BoundDeconstructionAssignmentOperator(node, FlattenDeconstructVariables(checkedVariables), boundRHS, deconstructions, assignments, voidType, hasErrors: hasErrors);
             }
             finally
```
`try/finally` seems unnecessary here.

jcouv replied:
You're right. Removed.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -132,23 +147,74 @@ private BoundExpression BindDeconstructionAssignment(AssignmentExpressionSyntax
+            for (int i = 0; i < varCount; i++)
+            {
+                var variable = variables[i];
+                if (!variable.IsNested && variable.Single.Kind == BoundKind.DeconstructionLocalPendingInference)
```
This same expression is used in the lambda above. Consider extracting a helper method.

jcouv replied:
Done.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -132,23 +147,74 @@ private BoundExpression BindDeconstructionAssignment(AssignmentExpressionSyntax
+                var variable = variables[i];
+                if (!variable.IsNested && variable.Single.Kind == BoundKind.DeconstructionLocalPendingInference)
+                {
+                    if (i < foundCount)
```
Loop until `matchCount` instead?

jcouv replied:
Yes. Simplified.

gafter commented:
> Do you see any concerns if that new node simply packages a `BoundDeconstructionAssignmentOperator` (rather than replicating its fields)?

If you add a field to the new node containing the list of declared variables, I think that is a good solution.

AlekseyTs commented:
I think all tests for various forms of deconstruction declaration should also test behavior of SemanticModel.GetDeclaredSymbol API for all declared locals.

gafter commented:
You will want to test the region analysis APIs for these new variables, but doing that in a separate PR makes sense.

jcouv commented:
Added semantic model API tests and the initial feedback. 
I have not changed the statement node yet though (that will come shortly).
I took a note to test region analysis API in #11299.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -15,20 +16,43 @@ namespace Microsoft.CodeAnalysis.CSharp
+            return result;
+        }
+
+        private void FreeDeconstructionVariables(ArrayBuilder<DeconstructionVariable> variables)
```
`static`

jcouv replied:
Done

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -132,23 +156,73 @@ private BoundExpression BindDeconstructionAssignment(AssignmentExpressionSyntax
+            if (!variables.Any(v => IsDeconstructionLocalPendingInference(v)))
+            {
+                return;
+            }
```
Is this check necessary? It doesn't look like we'd do any additional work in the loop below in that case.

cston replied:
Is this check necessary? It doesn't look like we'd do any additional work in the loop below in that case.

And if the check is removed, `IsDeconstructionLocalPendingInference` could probably be inlined in the one remaining caller. (Sorry for not noticing that earlier.)

jcouv replied:
You're right. Previously I had to rebuild an `ImmutableArray` and so this shortcut made sense. I'll remove.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -132,23 +156,73 @@ private BoundExpression BindDeconstructionAssignment(AssignmentExpressionSyntax
+        /// <summary>
+        /// Find any deconstruction locals that are still pending inference and fail their inference.
+        /// </summary>
+        private void FailRemainingInferences(ArrayBuilder<DeconstructionVariable> variables, DiagnosticBag diagnostics)
```
`static`.

jcouv replied:
Done

jcouv replied:
This method actually uses `this`.

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -405,13 +481,98 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+                else
+                {
+                    variablesBuilder.Add(new DeconstructionVariable(BindDeconstructionDeclarationVariables(variable, varSyntax, diagnostics)));
+                }
```
Consider moving condition inside `variablesBuilder.Add(...)`, or move `variablesBuilder.Add()` outside of `if`.

jcouv replied:
Done

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -549,7 +549,14 @@ private BoundStatement BindLocalDeclarationStatement(LocalDeclarationStatementSy
+            else
+            {
+                return binder.WrapWithVariablesIfAny(node, binder.BindDeconstructionDeclarationStatementParts(node, diagnostics));
+            }
```
Consider moving condition inside `binder.WrapWithVariablesIfAny(...)` or move `binder.WrapWithVariablesIfAny()` outside of `if`.

jcouv replied:
Done.

cston commented on src/Compilers/CSharp/Portable/BoundTree/Expression.cs:
```diff
@@ -2886,4 +2886,22 @@ public override void Accept(OperationVisitor visitor)
+    {
+        public override void Accept(OperationVisitor visitor)
+        {
+            visitor.VisitNoneOperation(this);
```
`throw ExceptionUtilities.Unreachable;` here and below.

AlekseyTs replied:
That would make sense since we don't expect this node to survive initial binding.

---
In reply to: [68700093](https://github.com/dotnet/roslyn/pull/12224#discussion_r68700093) [](ancestors = 68700093)

jcouv replied:
Done.
`OutVarLocalPendingInference` does implement OperatiionKind.None however, even though it also does not survive initial binding. Should it be changed too?

AlekseyTs replied:
If you could, it would be great.

---
In reply to: [68797293](https://github.com/dotnet/roslyn/pull/12224#discussion_r68797293) [](ancestors = 68797293)

jcouv replied:
Done

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -580,5 +598,58 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
+                    topLevelVariableDeclaration = (VariableDeclarationSyntax)grandParent;
+                }
+
+                Debug.Assert(topLevelVariableDeclaration.Deconstruction != null && topLevelVariableDeclaration.Deconstruction.Value != null);
```
Consider splitting `&&` into two asserts so it's clear which condition was false when an assert fails.

jcouv replied:
Done.

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -580,5 +598,58 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
+
+                var topLevelVariableDeclaration = (VariableDeclarationSyntax)IdentifierToken.
+                                                                    Parent. // VariableDeclaratorSyntax
+                                                                    Parent; // VariableDeclarationSyntax
```
Minor point: casts to `VariableDeclarationSyntax` are unnecessary.

jcouv replied:
Yes. Removed them from here and added them in the asserts instead.

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -580,5 +598,58 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
+                Debug.Assert(topLevelVariableDeclaration.Deconstruction != null && topLevelVariableDeclaration.Deconstruction.Value != null);
+
+                var statement = topLevelVariableDeclaration.Parent;
+                TypeSymbol result;
```
If `result` is necessary, move into `case`.

jcouv replied:
Done

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -580,5 +598,58 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
+
+                Debug.Assert(topLevelVariableDeclaration.Deconstruction != null && topLevelVariableDeclaration.Deconstruction.Value != null);
+
+                var statement = topLevelVariableDeclaration.Parent;
```
Could the loop above be simplified to walk up the `Parent` chain from `IdentifierToken` to the nearest `LocalDeclarationStatementSyntax`? Then the `switch` below could be removed as well.

jcouv replied:
Let me take a note but hold off on this. I think the switch case will be expanded for 'for' and 'foreach' and other contexts for such locals.

jaredpar commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -132,23 +156,73 @@ private BoundExpression BindDeconstructionAssignment(AssignmentExpressionSyntax
+        /// <summary>
+        /// Inform the variables about found types (whether one was found or not).
+        /// </summary>
+        static private void SetInferredTypes(ArrayBuilder<DeconstructionVariable> variables, ImmutableArray<TypeSymbol> foundTypes)
```
NIT: generally `private static`

jcouv replied:
Fixed here and other instances I introduced.

jaredpar commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -132,23 +156,73 @@ private BoundExpression BindDeconstructionAssignment(AssignmentExpressionSyntax
+                return;
+            }
+
+            var matchCount = Math.Min(variables.Count, foundTypes.Length);
```
In what case would we have more found types than variables?

jcouv replied:
`var (x, y) = (1, 2, 3);` would.

jaredpar replied:
Gotcha.  In that particular case where is the error generated for the mismatched counts? 

---
In reply to: [68799374](https://github.com/dotnet/roslyn/pull/12224#discussion_r68799374) [](ancestors = 68799374)

jcouv replied:
`MakeTupleDeconstructStep` on line 131. It first calls `SetInferredTypes` to record the types it found, then it checks for matching cardinality.

jaredpar commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -158,21 +232,21 @@ private BoundExpression BindDeconstructionAssignment(AssignmentExpressionSyntax
             }
 
-            public bool IsNested => !Nested.IsDefault;
+            public bool IsNested => Nested != null;
```
My reaction to `IsNested` is to think "is this variable nested".  In this case it seems to mean that it's a variable that has nested declarations.  I would assume this variable `HasNestedDeclarations` and those  variables would be `IsNested`.  

Am I missing something here?

jcouv replied:
You're right, the name of this property could be clearer. How about `HasNestedVariables`?
`HasNestedDeclarations` is not so good because this structure is also used for d-assignment (which doesn't declare variables, just references them).

jaredpar replied:
`HasNestedVariables` sounds good to me.

---
In reply to: [68803104](https://github.com/dotnet/roslyn/pull/12224#discussion_r68803104) [](ancestors = 68803104)

jaredpar commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -405,13 +481,98 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+            Debug.Assert(node.Deconstruction != null);
+            var variables = node.Deconstruction.Variables;
+
+            if (varSyntax == null)
```
Wouldn't it be clearer to force `varSyntax` to be non null and have callers explicitly pass down the type?  At the call site `node.Type` reads better than `null`

jcouv replied:
`varSyntax` is expected to be null in cases like `(int x, int y) = M();` which have no var.

jcouv replied:
I'm thinking about this more.

jcouv replied:
I moved this logic to the call-site as you suggested.

jaredpar commented on src/Compilers/CSharp/Portable/BoundTree/DeconstructionLocalPendingInference.cs:
```diff
@@ -0,0 +1,29 @@
+{
+    internal partial class DeconstructionLocalPendingInference
+    {
+        public BoundLocal SetInferredType(TypeSymbol type, bool success)
```
In the case of `success` should we assert that `type` is non-null
In the case of `!success` do we have any guarantese about `type`?  Is it null, error, etc ....

jcouv replied:
The only `!success` case is below and passes a non-null type as well. I'll add an assert.

jaredpar commented on src/Compilers/CSharp/Portable/Errors/ErrorCode.cs:
```diff
@@ -1407,5 +1407,6 @@ internal enum ErrorCode
         ERR_DeconstructTooFewElements = 8213,
 
         ERR_ConversionNotTupleCompatible = 8214,
+        ERR_TypeInferenceFailedForImplicitlyTypedDeconstructionLocal = 8215,
```
Thought this might be the longest error message identifier but nope seems there are a few that are longer

jcouv replied:
Hehe. 
Shortened to `ERR_TypeInferenceFailedForDeconstructionLocal` as type inference could only fail if the type is implicit...

AlekseyTs commented:
The test failures are might be related to the change.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -77,7 +77,7 @@
 
   <AbstractNode Name="BoundValuePlaceholderBase" Base="BoundExpression">
-    <Field Name="Type" Type="TypeSymbol" Override="true" Null="disallow"/>
+    <Field Name="Type" Type="TypeSymbol" Override="true"/>
```
Removing ```Null="disallow"``` feels suspicious, the whole point of placeholders is to have type. However, I am at the very beginning of the review process, I might change opinion as I make more progress.

AlekseyTs replied:
I didn't find any place where we take advantage of this change. Revert?

---
In reply to: [68790405](https://github.com/dotnet/roslyn/pull/12224#discussion_r68790405) [](ancestors = 68790405)

jcouv replied:
Reverted indeed. Thanks!

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -580,5 +598,58 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
+                while (true)
+                {
+                    var parent = topLevelVariableDeclaration.Parent;
+                    if (parent.Kind() != SyntaxKind.VariableDeconstructionDeclarator) { break; }
```
Each ```{```/```}``` should be on a separate line.

AlekseyTs replied:
BTW, is there a guarantee that parent is not null?

---
In reply to: [68792597](https://github.com/dotnet/roslyn/pull/12224#discussion_r68792597) [](ancestors = 68792597)

jcouv replied:
Fixed the formatting for braces.

Good question on null parent. I only create such local symbols when I am in a valid context. Do you think I should have additional guards?
If so, does `PossiblyImplicitlyTypedOutVarLocalSymbol` have the same problem?

AlekseyTs replied:
I think guards never hurt. I believe the shape of the syntax used by PossiblyImplicitlyTypedOutVarLocalSymbol is verified elsewhere before instances are created. 

---
In reply to: [68805575](https://github.com/dotnet/roslyn/pull/12224#discussion_r68805575) [](ancestors = 68805575)

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -580,5 +598,58 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
+                switch (statement.Kind())
+                {
+                    case SyntaxKind.LocalDeclarationStatement:
+                        this.binder.BindDeconstructionDeclarationStatementParts((LocalDeclarationStatementSyntax)statement, diagnostics);
```
This might be not the right binder to use, this is the binder that owns the scope for this local, but the local declaration statement has its own scope for pattern and out var locals. That binder should be in the chain at this point. Try using out vars and pattern vars in the expression part of the declaration. It looks like we have similar problem with LocalWithInitializer and other similar C# 6 types. I suggest to not try to be smart here and simply call the method that binds the whole declaration statement, it should locate the right binder for the job.

jcouv replied:
I see. I'll call `BindLocalDeclarationStatement` instead (it calls `GetBinder(node)`).
I didn't understand the problem you pointed out with `LocalWithInitializer`. Should I file an issue to follow up?

AlekseyTs replied:
> I didn't understand the problem you pointed out with `LocalWithInitializer`. Should I file an issue to follow up?

No, I'll do it myself. Thanks. 

---
In reply to: [68823222](https://github.com/dotnet/roslyn/pull/12224#discussion_r68823222) [](ancestors = 68823222)

AlekseyTs replied:
Opened [https://github.com/dotnet/roslyn/issues/12266](https://github.com/dotnet/roslyn/issues/12266)

---
In reply to: [68833592](https://github.com/dotnet/roslyn/pull/12224#discussion_r68833592) [](ancestors = 68833592,68823222)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/LocalScopeBinder.cs:
```diff
@@ -181,6 +189,34 @@ protected ImmutableArray<LocalSymbol> BuildLocals(SyntaxList<StatementSyntax> st
+            else
+            {
+                Debug.Assert(declaration.Variables.Count == 1);
+                var declarator = declaration.Variables[0];
```
Please spell out the type.

jcouv replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/LocalScopeBinder.cs:
```diff
@@ -181,6 +189,34 @@ protected ImmutableArray<LocalSymbol> BuildLocals(SyntaxList<StatementSyntax> st
+                                                            this,
+                                                            declaration.Type,
+                                                            declarator.Identifier,
+                                                            isVar: varSyntax != null);
```
Are we judging about the type based on ```varSyntax```, or ```declaration.Type```? I think we should pick one or the other. Also, whether the type is ```var``` should be determined based on content of the type syntax and MakeDeconstructionLocal should be able to do this on its own, without an extra parameter.

jcouv replied:
The `var` syntax can come from multiple parents up. For example, for `y` in `var (x, (y, z))`. So that is why I prefer to pass it in.
Two examples:

1. In `var (x, (y, z))`, we create a local that may be implicitly typed for `y` because isVar will be true. That local will have its type syntax set to null.
2. In `(var x, int y)`, we should also create an implicitly typed local for `x`, with type syntax pointing to `var`. I think that's a bug there. I'll fix.

AlekseyTs replied:
I think the closest type syntax should win. If declaration.Type is not null, it should be used, otherwise varSyntax should be used. But we should still check if varSyntax is actually a var syntax.

---
In reply to: [68807675](https://github.com/dotnet/roslyn/pull/12224#discussion_r68807675) [](ancestors = 68807675)

jcouv replied:
I'm ok to make that change, but it cannot be observed. No types (including `var`) may appear once you enter a var form.

AlekseyTs replied:
> it cannot be observed

Even if that is the case today. Judging about different "properties" of the type based on the same syntax makes code easier to reason about, less error prone, and more robust against future changes. For example we might relax parser behavior in error recovery to allow types here, even though it would cause parse errors.

---
In reply to: [68824220](https://github.com/dotnet/roslyn/pull/12224#discussion_r68824220) [](ancestors = 68824220)

jcouv replied:
Makes sense. Thanks

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:
```diff
@@ -549,7 +549,14 @@ private BoundStatement BindLocalDeclarationStatement(LocalDeclarationStatementSy
             var binder = GetBinder(node);
```
This is the place that ensures the right binder is used (related to my comment in PossiblyImplicitlyTypedDeconstructionLocalSymbol class).

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -132,23 +156,73 @@ private BoundExpression BindDeconstructionAssignment(AssignmentExpressionSyntax
+                var variable = variables[i];
+                if (IsDeconstructionLocalPendingInference(variable))
+                {
+                    var local = ((DeconstructionLocalPendingInference)variable.Single).SetInferredType(foundTypes[i], true);
```
Consider using named argument for ```true```.

AlekseyTs replied:
Please spell out the type.

---
In reply to: [68801612](https://github.com/dotnet/roslyn/pull/12224#discussion_r68801612) [](ancestors = 68801612)

AlekseyTs replied:
Please spell out the type.

jcouv replied:
Done for both.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -132,23 +156,73 @@ private BoundExpression BindDeconstructionAssignment(AssignmentExpressionSyntax
+                {
+                    if (variable.Single.Kind == BoundKind.DeconstructionLocalPendingInference)
+                    {
+                        var local = ((DeconstructionLocalPendingInference)variable.Single).FailInference(this, diagnostics);
```
Do we ever want to report errors here? For example, it feels like when we call this method from BindDeconstructionAssignment, we don't want to report errors here since we already complained about real problems.

jcouv replied:
The primary purpose for failing remaining inferences is to eliminate any remaining instances of `DeconstructionLocalPendingInference`. You're right I don't care as much about reporting additional errors.
I'll remove the error reporting inside `FailInference` to avoid duplicates (as seen in the last test, called `Mismatch`).

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -405,13 +481,98 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+            // - type and declarators are set, but deconstruction is null. This could represent `int x`, which is a single variable.
+            // - type is null, declarators are set, but deconstruction is null. This could represent `x`, which is a single variable.
+            // - type is set to 'var', declarators are null, and deconstruction is set. This could represent `var (...)`
+            // - type and declarators are null, but deconstruction is set. Thi could represent `(int x, ...)`
```
Typo ```Thi```?

jcouv replied:
Fixed

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -405,13 +481,98 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+        private ArrayBuilder<DeconstructionVariable> BindDeconstructionDeclarationVariables(VariableDeclarationSyntax node, TypeSyntax varSyntax, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.Deconstruction != null);
+            var variables = node.Deconstruction.Variables;
```
Please spell out the type. ```var``` can be used only if the type is spelled out in the initializer.

jcouv replied:
Sorry about that. Fixed

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -405,13 +481,98 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+            Debug.Assert(node.Variables.Count == 1);
+
+            var declarator = node.Variables[0];
+            var typeSyntax = varSyntax ?? node.Type;
```
I would expect the more local type (node.Type) to take the precedence.

jcouv replied:
No. If you have a `var` on the outside, then you cannot have an explicit type on a nested variable.
This is ok: `var (x, (y, z))`.
But this is not: `var (x, (int y, string z))`.

AlekseyTs replied:
I know it is an error, but still using the closes non-null type syntax, even in error recovery case is, better I think. In the example that you gave, type of y should be int and type of z should be string, we shouldn't try to infer them. And don't forget that 'var' isn't var until proven so. So, 'var (x, (y, z))' could be invalid too and the type could actually be explicit in this case. BTW, do we have a unit-test for a scenario like that?

---
In reply to: [68813077](https://github.com/dotnet/roslyn/pull/12224#discussion_r68813077) [](ancestors = 68813077)

jcouv replied:
Renamed argument to `closestTypeSyntax`.
I didn't realize I was computing it twice (here and in caller). I'll simplify. Thanks

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -405,13 +481,98 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+            AliasSymbol alias;
+            bool isVar;
+            TypeSymbol declType = BindVariableType(node, diagnostics, typeSyntax, ref isConst, out isVar, out alias);
+            Debug.Assert(isVar || varSyntax == null); // if varSyntax is set, then isVar is true
```
I don't think this assert is correct. 'var' can still bind to a real type. Please add a test like this.

jcouv replied:
I think I have one:

``` C#
        public void VarType()
        {
            string source = @"
class C
{
    static void Main()
    {
        (var x1, int x2) = (new var(), 2);
        System.Console.WriteLine(x1 + "" "" + x2);
    }
}
class var
{
    public override string ToString() { return ""var""; }
}
";
            var comp = CompileAndVerify(source, expectedOutput: "var 2 ", additionalRefs: new[] { ValueTupleRef, SystemRuntimeFacadeRef }, parseOptions: TestOptions.Regular);
            comp.VerifyDiagnostics();
        }
```

jcouv replied:
I'm not sure which fix caused the change, but I am now hitting that assert ;-)
I'll remove it.

AlekseyTs replied:
I was actually thinking more about "var (x, y) = ..." where var turned out to be a real type.

---
In reply to: [68813504](https://github.com/dotnet/roslyn/pull/12224#discussion_r68813504) [](ancestors = 68813504)

jcouv replied:
I'll add a test for that.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -405,13 +481,98 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+            // Check for variable declaration errors.
+            // Use the binder that owns the scope for the local because this (the current) binder
+            // might own nested scope.
+            bool hasErrors = localSymbol.Binder.ValidateDeclarationNameConflictsInScope(localSymbol, diagnostics);
```
Do we have tests for errors reported by this function?

jcouv replied:
I'm not sure. Let me check.

jcouv replied:
I checked that I do have a test to hit the error case with this function. 
See `DeclareLocalTwice` which tests `var (x1, x1) = ...`.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -405,13 +481,98 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+            }
+            else
+            {
+                DeclareLocalVariable(localSymbol, declarator.Identifier, declType);
```
Is this call necessary in a non-var case?

jcouv replied:
No, indeed. Removed

AlekseyTs commented:
Should test SemanticModel.GetSymbolInfor, SemanticModel.GetTypeInfo for all type syntax nodes withing deconstruction declaration.

AlekseyTs commented:
I am done with review pass.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -580,5 +598,58 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
+                    if (parent.Kind() != SyntaxKind.VariableDeconstructionDeclarator) { break; }
+
+                    var grandParent = parent.Parent;
+                    if (grandParent.Kind() != SyntaxKind.VariableDeclaration) { break; }
```
Same comments here.

AlekseyTs commented:
Also should test SemanticModel.GetAliasInfo

jcouv commented:
I pushed a commit with most feedback addressed, except: I need to add tests for errors from `ValidateDeclarationNameConflictsInScope` and tests for additional semantic model APIs that Aleksey suggested.
Also, I'm still investigating bug with getting the type symbol for "var" in `(var x, var y) =...`

cston commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -580,5 +600,86 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
+                Debug.Assert(((VariableDeclarationSyntax)topLevelVariableDeclaration).Deconstruction != null);
+                Debug.Assert(((VariableDeclarationSyntax)topLevelVariableDeclaration).Deconstruction.Value != null);
+
+                var statement = topLevelVariableDeclaration.Parent;
```
Perhaps extract a `GetLocalDeclarationStatement(IdentifierToken)` helper method that is used here and in the `.ctor` above and have that helper method assert there at least one pair of { `VariableDeconstructionDeclarator`, `VariableDeclaration` }.

jcouv replied:
Refactored all this code to minimize duplication.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs:
```diff
@@ -1735,6 +1735,11 @@ public override BoundNode VisitDeconstructionAssignmentOperator(BoundDeconstruct
             return null;
         }
 
+        public override BoundNode VisitLocalDeconstructionDeclaration(BoundLocalDeconstructionDeclaration node)
```
Do we need to override this method? Would base implementation work?

jcouv replied:
Indeed, I can remove this. Thanks

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ExpressionStatement.cs:
```diff
@@ -26,6 +26,23 @@ public override BoundNode VisitExpressionStatement(BoundExpressionStatement node
             }
         }
 
+        public override BoundNode VisitLocalDeconstructionDeclaration(BoundLocalDeconstructionDeclaration node)
```
Would LocalRewriter_LocalDeclaration.cs be a better location for this function?

jcouv replied:
The reason I have it here is that it follows from `VisitExpressionStatement` above. Unfortunately, I didn't find a good way to factor the common logic.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:
```diff
@@ -2586,42 +2596,557 @@ class C
+            var symbol = model.GetDeclaredSymbol(variableDeclaratorSyntax);
+            Assert.Equal(variableDeclaratorSyntax.Identifier.ValueText, symbol.Name);
+            Assert.Equal(LocalDeclarationKind.RegularVariable, ((LocalSymbol)symbol).DeclarationKind);
+            Assert.Same(symbol, model.LookupSymbols(decl.SpanStart, name: variableDeclaratorSyntax.Identifier.ValueText).Single());
```
Why is the previous assert removed?

AlekseyTs replied:
I believe it calls a different overload of GetDeclaraedSymbol.

---
In reply to: [68856288](https://github.com/dotnet/roslyn/pull/12224#discussion_r68856288) [](ancestors = 68856288)

jcouv replied:
I didn't realize it was a different overload. I thought it became the same call as a few lines above. 
Restored it.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -83,6 +83,26 @@ internal Binder Binder
+        {
+            Debug.Assert(parentTypeSyntax == null || parentTypeSyntax.IsVar);
+
+            if (parentTypeSyntax != null || typeSyntax.IsVar)
```
We are still dealing with two type syntaxes here. I think this function should take just one. The caller should decide which node to pass to this function, but it should be one node. We are always passing typeSyntax to the local symbol. Could it be null? In cases when it is not null and not var, is there a reason to create PossiblyImplicitlyTypedDeconstructionLocalSymbol?

jcouv replied:
Switched to a single parameter (closestTypeSyntax).

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceLocalSymbol.cs:
```diff
@@ -580,5 +600,86 @@ protected override TypeSymbol InferTypeOfVarVariable(DiagnosticBag diagnostics)
+                while (true)
+                {
+                    var parent = topLevelVariableDeclaration.Parent;
+                    if (parent.Kind() != SyntaxKind.VariableDeconstructionDeclarator)
```
Should we still check parent for null? For out variables, PatternVariableFinder enforces the shape before calling MakeLocal.

jcouv replied:
I refactored this code to be safe with nulls and in a re-usable method.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/LocalScopeBinder.cs:
```diff
@@ -181,6 +189,34 @@ protected ImmutableArray<LocalSymbol> BuildLocals(SyntaxList<StatementSyntax> st
+                SourceLocalSymbol localSymbol = SourceLocalSymbol.MakeDeconstructionLocal(
+                                                            this.ContainingMemberOrLambda,
+                                                            this,
+                                                            declaration.Type,
```
I think we should simply pass ```declaration.Type ?? parentTypeSyntax``` here and get rid of the last argument.

jcouv replied:
I see what you mean. That'll work. I'll change.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -405,13 +471,98 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+            var localsBuilder = ArrayBuilder<DeconstructionVariable>.GetInstance(variables.Count);
+            foreach (var variable in variables)
+            {
+                TypeSyntax variableVarSyntax = varSyntax ?? variable.Type;
```
I thin for consistency with other parts of the compiler this should be ```variable.Type ?? varSyntax```

jcouv replied:
Thanks. Fixed.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -405,13 +471,98 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+            Debug.Assert(node.Variables.Count == 1);
+
+            var declarator = node.Variables[0];
+            var typeSyntax = varSyntax ?? node.Type;
```
BindDeconstructionDeclarationVariables already passed type syntax to us in varSyntax parameter (the name is misleading, BTW), is there a reason to calculate it again?

jcouv replied:
Seems like a comment was lost. I removed the double-calculation and renamed the parameter to `closestTypeSyntax` (here and some other places).

cston commented on src/Compilers/CSharp/Portable/Syntax/SyntaxFacts.cs:
```diff
@@ -381,5 +381,65 @@ internal static bool IsVar(this Syntax.InternalSyntax.SyntaxToken node)
+            if (((VariableDeclarationSyntax)parent).Deconstruction != null)
+            {
+                return true;
+            }
```
```return ((VariableDeclarationSyntax)parent).Deconstruction != null;```

jcouv replied:
Fixed

cston commented:
LGTM

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/Expression.cs:
```diff
@@ -2870,20 +2885,46 @@ public override void Accept(OperationVisitor visitor)
         public override void Accept(OperationVisitor visitor)
         {
-            visitor.VisitNoneOperation(this);
+            throw ExceptionUtilities.Unreachable;
```
Thanks!

AlekseyTs commented on src/Compilers/CSharp/Portable/Syntax/SyntaxFacts.cs:
```diff
@@ -381,5 +381,65 @@ internal static bool IsVar(this Syntax.InternalSyntax.SyntaxToken node)
+        /// Figures out if this expression is a type in a deconstruction-declaration.
+        /// Outputs the VariableDeclarationSyntax directly containing it, if that is the case.
+        /// </summary>
+        internal static bool IsDeconstructionType(ExpressionSyntax expression, out SyntaxNode parent)
```
```VariableDeclarationSyntax parent```?

jcouv replied:
I had that and changed it back. It was making the call site more trouble. I'll take another look.

jcouv replied:
I tried this again and I think it makes the code worse. Need to hold more locals, more casts, and also call site needs another dedicated local.
I'll leave as-is.

AlekseyTs commented on src/Compilers/CSharp/Portable/Syntax/SyntaxFacts.cs:
```diff
@@ -381,5 +381,65 @@ internal static bool IsVar(this Syntax.InternalSyntax.SyntaxToken node)
+        /// Figures out if this token is an identifier in a deconstruction-declaration.
+        /// Outputs the top-level VariableDeclarationSyntax if that is the case.
+        /// </summary>
+        internal static bool IsDeconstructionIdentifier(SyntaxToken identifier, out SyntaxNode parent)
```
Same suggestion here.

AlekseyTs commented on src/Compilers/CSharp/Portable/Syntax/SyntaxFacts.cs:
```diff
@@ -381,5 +381,65 @@ internal static bool IsVar(this Syntax.InternalSyntax.SyntaxToken node)
+            return IsDeconstruction((VariableDeclarationSyntax)parent, out parent);
+        }
+
+        private static bool IsDeconstruction(VariableDeclarationSyntax declaration, out SyntaxNode parent)
```
Same suggestion here.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -405,13 +471,97 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+            // might own nested scope.
+            bool hasErrors = localSymbol.Binder.ValidateDeclarationNameConflictsInScope(localSymbol, diagnostics);
+
+            if (node.Type != null)
```
I am not sure why are we looking at ```node.Type``` here. I think we should just use ```closestTypeSyntax``` as the type syntax for the local and should remove this null check because we again looking at two different nodes. Should look just at one.

jcouv replied:
There are four cases, which can be thought of as 2x2.
On one dimension, is the "var" next to the identifier: `var (x, y)` vs. `(var x, ...)`.
On the other dimension, is there an available type called "var".

     | "var" type exists | or doesn't exist |
-----| ---------------------|----------------------|
var(x,y) |   inference needed | inference needed |
(var x, ...)  | no inference | inference needed|

Only if "var" is the current node's type (second row), then there is need to see if a "var" type exists to decide if the type will need inference.
If we check if the `closestTypeSyntax` is "var", then we would un-necessarily check whether a "var" type exists in the case on the top left corner.

AlekseyTs replied:
See my other response (about the test), in my opinion, we should never make judgement about 'var' just based on syntax. This is inconsistent with current handling of 'var' throughout the compiler.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -405,13 +471,97 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+                }
+            }
+
+            return new DeconstructionLocalPendingInference(declarator, localSymbol);
```
Why are we making an assumption that closestTypeSyntax is real var?

AlekseyTs commented on src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:
```diff
@@ -535,6 +535,18 @@ public SymbolInfo GetSymbolInfo(ExpressionSyntax expression, CancellationToken c
+                    return new SymbolInfo(deconstructionType);
+                }
+
+                return SymbolInfo.None;
```
It looks like this code is pretty much a duplication of the previous block. Consider extracting into a method and passing target declarator as a parameter.

jcouv replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:
```diff
@@ -2586,42 +2596,651 @@ class C
+    static void Main()
+    {
+        (var x1, int x2) = (new var(), 2);
+        var (x3, x4) = (3, new var());
```
Shouldn't this be an error? Is it an error for ```int (a, b) = ...```? Then I would say when ```var``` is a real type, ```var (a, b)``` should be an error. For example, what if ```var``` is just an alias ```using var = System.Int32;```? BTW, consider adding a test with alias as well.

jcouv replied:
See previous comment for more details, but my thinking was that no type is allowed in that position, so "var" is un-ambiguous and always means "type to be inferred".

Even if we wanted to allow types in that deconstruction form in the future, the most likely syntax that we discussed was `(int, string) (x1, x2) = ...`, not `int (x1, x2) = ...`, if I remember the discussion correctly. 

Thanks for the suggestion of alias tests in semantic model. I'll add. Another var ambiguity test on my list is having a ref-returning var method.

AlekseyTs replied:
```var``` is never unambiguous, it should always be bound first in order to determine if it is a designator for an implicit type. Some folks define a type named "var" with the intent of disabling implicit typing in their programs. I have a very strong felling that this scenario should be an error. If you have doubts about that, bring it do LDM.

jcouv replied:
Ok. I will report an error if a "var" type exists and you write `var (x, y) = ...`.
I'll take a note to confirm with LDM as well.

AlekseyTs commented:
Done with review pass.

jcouv commented:
Pushed a change with addresses the feedback and suggestions.

I will add more tests on semantic model (including aliases) as well as the `ValidateDeclarationNameConflictsInScope` error case in subsequent PR.

jcouv commented:
The last commit makes `var (x, y) = ...` an error if the `var` type exists. 
The diagnostics are not great (too many errors reported), but should be ok for this context.
